src/layer/AtmosphereLayer.js

```diff
@@ -0,0 +1,414 @@
+             */
+            skyWidth: {
+                get: function () {
+                    return this._skyWidth;
+                },
+                set: function (value) {
+                    this._skyWidth = value;
+                }
+            },
+
+            /**
+             * The number of latitudinal points in the grid.
+             * @memberof AtmosphereLayer.prototype
+             * @type {Number}
+             */
+            skyHeight: {
+                get: function () {
+                    return this._skyHeight;
+                },
+                set: function (value) {
+                    this._skyHeight = value;
+                }
+            },
+
+            /**
+             * The grid of sky Cartesian points.
+             * @memberof AtmosphereLayer.prototype
+             * @type {Float32Array}
+             */
+            skyPoints: {
+                get: function () {
+                    return this._skyPoints;
+                },
+                set: function (value) {
+                    this._skyPoints = value;
+                }
+            },
+
+            /**
+             * The triangle strip sky indices.
+             * @memberof AtmosphereLayer.prototype
+             * @type {Number[]}
+             */
+            skyTriStrip: {
+                get: function () {
+                    return this._skyTriStrip;
+                },
+                set: function (value) {
+                    this._skyTriStrip = value;
+                }
+            }
+        });
+
+        // Documented in superclass.
+        AtmosphereLayer.prototype.doRender = function (dc) {
+
+            if (dc.globe.is2D()) {
+                return;
+            }
+
+            this.determineLightDirection(dc);
+            this.drawSky(dc);
+            this.drawGround(dc);
+        };
+
+        // Internal. Intentionally not documented.
+        AtmosphereLayer.prototype.applySkyVertices = function (dc) {
+
+            var gl = dc.currentGlContext,
+                program = dc.currentProgram,
+                skyData = this.skyData,
+                vboId;
+
+            if (!skyData.verticesVboCacheKey) {
+                skyData.verticesVboCacheKey = dc.gpuResourceCache.generateCacheKey();
+            }
+
+            vboId = dc.gpuResourceCache.resourceForKey(skyData.verticesVboCacheKey);
+            if (!vboId) {
+                vboId = gl.createBuffer();
+                dc.gpuResourceCache.putResource(skyData.verticesVboCacheKey, vboId,
+                    this.skyPoints.length * 4);
+                skyData.refreshVertexBuffer = true;
+            }
+
+            gl.bindBuffer(gl.ARRAY_BUFFER, vboId);
+            if (skyData.refreshVertexBuffer) {
+                gl.bufferData(gl.ARRAY_BUFFER, new Float32Array(this.skyPoints), gl.STATIC_DRAW);
+                dc.frameStatistics.incrementVboLoadCount(1);
+                skyData.refreshVertexBuffer = false;
+            }
+
```

Comment:
I don't see where SkyProgram specifies the vertexPointLocation attribute. Additionally, the emerging pattern in WebWW is to specify vertex attribute locations explicitly at GLSL compilation time, then hard code the specified locations in our GL calls. In this case, program.vertexPointLocation can be replaced with 0.